### PR TITLE
Introduce __STACKTRACE__

### LINKED ISSUE #7097: Introduce __STACKTRACE__
**TL;DR** Replace the use of `System.get_stacktrace()` with a pseudo-variable `__STACKTRACE__` similar to `__MODULE__` or `__CALLER__` available only inside `catch` and `rescue` clauses of the `try` expression.

## Reasons
Erlang recently set out to overhaul how stack traces are handled. The `:erlang.get_stacktrace/0` call (and thus it's Elixir's equivalent `System.get_stacktrace/0`) are problematic - they may cause the emulator to hold on to the stack trace for a very long time. This can be harmful, since the stack trace can be very big - in case of `FunctionClauseError` it includes all the arguments. The emulator has to hold on the stack trace for the last exception in general until the next exception is raised - which can be a long time.
To rectify this potential performance issue, OTP 20 started warning on the use of `:erlang.get_stacktrace()` outside of the `catch` clause of a `try` expression (in Elixir this would be outside of `catch` and `rescue`). The plan was to make the call return an empty list in some future Erlang version when called outside of `catch`.
OTP 21 will introduce an even more restrictive mechanism, to understand it we need to understand the syntax for `catch` in Erlang:
```erlang
try
  ...
catch
  Kind:Reason ->
    ...
end
```
Where `Kind` can be one of `:exit`, `:error` and `:throw`. The proposed (and merged) enhancement replaces this with:
```erlang
try
  ...
catch
  Kind:Reason:Stack ->
    ...
end
```
Where the `Stack` variable would contain the stack-trace for the just-rescued exception. It is forbidden to pattern match on this variable or use an already-bound variable (which would be equivalent to pinning it in Elixir). It's now planned that some future Erlang release will always return empty lists from `:erlang.get_stacktrace()` and the only way to acquire the stacktrace will be with this syntax. This means we need a compatible Elixir solution.

## Proposal
To solve this, we propose introduction of a new pseudo-variable `__STACKTRACE__` that would be available inside `catch` and `rescue` and would return the stack trace of the exception being currently handled.
This means that this code:
```elixir
try do
  ...
rescue
  e ->
    ...
    reraise(e, System.get_stacktrace())
end
```
Would become:
```elixir
try do
  ...
rescue
  e ->
    ...
    reraise(e, __STACKTRACE__)
end
```

The `__STACKTRACE__` pseudo-variable would be only available inside `catch` and `resuce` parts of `try` - it should fail to compile if present in any other place, similar to how `__CALLER__` is not available outside of macros. 

Having this as this kind of pseudo-variable is also compatible with how this works in Erlang - since neither pattern matching nor pinning of the stack variable is allowed, we don't loose any features.

## Implementation

Since OTP 21 is not stable yet, and we also want to support multiple Erlang versions, a conditional compilation of this feature is required. 

### On VMs without new syntax

When compiling on systems without support for the new syntax, the `__STACKTRACE__` pseudo-variable should compile into a call to `:erlang.get_stacktrace/0`.
Additionally, the call to `:erlang.get_stacktrace/0` should happen as the very first thing in the block, assigning the result to some internal variable, which could be later substituted for the `__STACKTRACE__` expression. This provides a very important advantage compared to the current `System.get_stacktrace/0` call by protecting against the "wrong stack trace" bug, which is very easy to introduce accidentally:
```elixir
try do
  ...
catch
  kind, reason ->
    cleanup()
    report_error(kind, reason, System.get_stacktrace())
end

defp cleanup() do
  {:ok, this_may_throw()}
catch
  throw -> {:error, throw}
end
```
If the `this_may_throw()` operation does indeed throw, the `System.get_stacktrace/0` call would return the value for that throw, instead of the exception we've just caught, since it always returns the stack trace of the last exception. With the use of the `__STACKTRACE__` pseudo-variable, this would compile into something conceptually like:
```elixir
try do
  ...
catch
  kind, reason ->
    stracktrace = :erlag.get_stacktrace()
    cleanup()
    report_error(kind, reason, stacktrace)
end
```
Which does not suffer from the issue. This means not only is the proposed feature giving us compatibility with Erlang, it is also a significant improvement in terms of preventing strange and hard to find bugs.

### On VMs with new syntax

When compiling on a VM that supports the new syntax, the `__STACKTRACE__` call should just use it, the previous example would compile into the following equivalent Erlang code:
```erlang
try do
  ...
catch
  Kind:Reason:Stacktrace ->
    cleanup()
    report_error(Kind, Reason, Stacktrace)
end
```

## Alternatives

An alternative would be to also support something like:
```elixir
try do
  ...
catch
  kind, reason, stack ->
    ...
end
```
There are couple of problems with that approach:
* similar to Erlang, it introduces this special variable that can't be pattern matched
* there's no clear way of using this syntax in combination with `rescue` or `catch` with just one pattern (which catches only throws).
* the compilation for VMs that don't support the new syntax is more awkward.

## Compatibility
Because of the conditional-compilation, the feature becomes compatible with all the Erlang releases. There is the constraint that code compiled on VM that supports the feature wouldn't work on older VMs, but this is also true today - code compiled on OTP 20 wouldn't work on OTP 18, even though Elixir in itself supports running on both VMs.

## Roadmap
The `System.get_stacktrace()` call should be soft-deprecated in the next release that includes the `__STACKTRACE__` feature. Additionally a loud deprecation should happen after 2 releases or as soon as Elixir declares compatibility for an Erlang VM that would always return an empty list for `:erlang.get_stacktrace()`. This might break the general deprecation policy approach of minimum 2 releases, but this is somewhat out of our hands.

### PR DESCRIPTION
**Pull request by michalmuskala**
This still needs probably more tests and additional docs, but I wanted to see if the approach for the implementation makes sense.

Closes #7097

### CODE REVIEW

File: lib/elixir/lib/kernel/special_forms.ex [hunk 1]
```diff
@@ -670,6 +670,14 @@ defmodule Kernel.SpecialForms do
   """
   defmacro __CALLER__, do: error!([])
 
+  @doc """
+  Returns the stacktrace for the curently handled exception.
+
+  It is available only in the `catch` and `rescue` clauses of `try/1`
+  expression.
```
Comment (whatyouhide) [inline lib/elixir/lib/kernel/special_forms.ex]: expression**s**

File: lib/elixir/src/elixir_erl_pass.erl [hunk 2]
```diff
@@ -629,3 +634,6 @@ generate_struct_name_guard([{map_field_exact, Ann, {atom, _, '__struct__'} = Key
   {lists:reverse(Acc, [{map_field_exact, Ann, Key, Match} | Rest]), S2};
 generate_struct_name_guard([Field | Rest], Acc, S) ->
   generate_struct_name_guard(Rest, [Field | Acc], S).
+
+format_error(invalid_stacktrace) ->
+  "__STACKTRACE__ is available only inside catch and rescue clauses of the try expression".
```
Comment (whatyouhide) [inline lib/elixir/src/elixir_erl_pass.erl]: "inside **the** catch and rescue clauses", or "inside catch and rescue clauses of the try expression**s**"?

File: lib/elixir/src/elixir_erl_pass.erl [hunk 3]
```diff
@@ -66,6 +66,11 @@ translate({'__block__', Meta, Args}, S) when is_list(Args) ->
 translate({'__CALLER__', Meta, Atom}, S) when is_atom(Atom) ->
   {{var, ?ann(Meta), '__CALLER__'}, S#elixir_erl{caller=true}};
 
+translate({'__STACKTRACE__', Meta, Atom}, S = #elixir_erl{stacktrace={Var, _}}) when is_atom(Atom) ->
+  {{var, ?ann(Meta), Var}, S#elixir_erl{stacktrace={Var, true}}};
+translate({'__STACKTRACE__', Meta, Atom}, S) when is_atom(Atom) ->
+  elixir_errors:form_error(Meta, S#elixir_erl.file, ?MODULE, invalid_stacktrace);
```
Comment (josevalim) [inline lib/elixir/src/elixir_erl_pass.erl]: We cannot raise errors inside elixir_erl pass. All validation should happen during expansion. We even removed the `.file` field in master. :)

**Comment by josevalim** [comment]
@michalmuskala sorry for the delay but I finally got to review this. :)

**Comment by michalmuskala** [comment]
@josevalim any ideas how to do this during the expansion pass? Is it ok to add another field to the `Macro.Env` struct? Or maybe we could reuse some place to store this information?

**Comment by josevalim** [comment]
@michalmuskala I would introduce a field called contextual_vars in Macro.Env that is a list of fields allowed in a given place. So once you enter catch/rescue, you would append `'__STACKTRACE__'`. When the variable is accessed, you check if the entry exists in the field.

This will also allow us to properly validate `__CALLER__`. Today it raises an awkward `{:unbound_var, var}` Erlang error.

**Comment by michalmuskala** [comment]
@josevalim My current idea is to generate new variable name before entering try and save it in the environment struct as you proposed. Later when expanding the try, we can use the saved variable (and mark it as used) and then annotate the `try` metadata that we use the stacktrace, so code generation in the erlang pass can do the right thing (we could do it in Elixir right now, but we won't be able when we start using the new `catch` argument).  I'm not clear, though, how could I generate a new variable name. In the erlang pass I could use the `elixir_erl_var:build/2`, but I don't see anything similar to use during the elixir pass.

**Comment by josevalim** [comment]
@michalmuskala I can think of two approaches.

In both cases, I don't think we should generate a new variable name in the Elixir pass.

### contextual_vars as a list

The `contextual_vars` simply says the given variable is allowed at that moment.

The generating of var would happen exclusively on the Erlang side, especially because we will ned to generate different types of code depending on the Erlang version.

### contextual_vars as a map

The `contextual_vars` is a map with the variable name as keys and if it was used or not as a boolean. If the key does not exist in the map, it means the variable is not available in that scope.

In this case, we could propagate the handling of the variable or not as a third argument in catch and a second argument in rescue. The con, in my opinion, is that we are changing the AST representation between passes and I would like to avoid those as much as possible.

Similar trade-offs happen in the metadata. Adding stuff to metadata makes it harder to write future passes. Remember we want the expanded AST to be translatable to JavaScript too.

### Wrapping up

I would go with keeping the `contextual_vars` as a list and simply validate the AST in the Elixir pass. The Erlang pass can be simpler because it will work on the validate AST, so it doesn't need to track context, only if stacktrace was used or not.

**Comment by michalmuskala** [comment]
Updated. I took the second approach making `contextual_vars` a list. Indeed handling all the variable generation on Erlang side makes much more sense.

File: lib/elixir/src/elixir_erl_pass.erl [hunk 4]
```diff
@@ -66,6 +66,9 @@ translate({'__block__', Meta, Args}, S) when is_list(Args) ->
 translate({'__CALLER__', Meta, Atom}, S) when is_atom(Atom) ->
   {{var, ?ann(Meta), '__CALLER__'}, S#elixir_erl{caller=true}};
 
+translate({'__STACKTRACE__', Meta, Atom}, S = #elixir_erl{stacktrace={Var, _}}) when is_atom(Atom) ->
+  {{var, ?ann(Meta), Var}, S#elixir_erl{stacktrace={Var, true}}};
```
Comment (josevalim) [inline lib/elixir/src/elixir_erl_pass.erl]: I think we can do like caller and simply use the `__STACKTRACE__` variable. This way the stacktrace field only keeps a boolean.

**Comment by josevalim** [comment]
Do you want to add `__CALLER__` handling here too or should we do it separately?

File: lib/elixir/src/elixir_erl_pass.erl (hunk 4 above)
Comment (michalmuskala) [reply lib/elixir/src/elixir_erl_pass.erl]: We can't since `try` (and stacktraces) could be nested, unlike `__CALLER__`.

**Comment by michalmuskala** [comment]
I wanted to do `__CALLER__` separately

**Comment by michalmuskala** [comment]
It seems there's some error on OTP 19, locally running `make test_elixir` just hangs...

**Comment by josevalim** [comment]
The 19 error has been fixed on master. :)
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**